### PR TITLE
Regenerate LibCURL.jl via Clang.jl

### LINKED ISSUE #52: Regenerate constants
Since the Clang-generated wrapper was generated 6 years ago, newer options such as `CURLOPT_UNIX_SOCKET_PATH` are missing. The constants, at least, should be regenerated.

### PR DESCRIPTION
**Pull request by nicoleepp**
The code used to regenerate is included in `gen/generate.jl`. Used Clang.jl on 0.7.

Closes #52 

LibCURL and FTPClient tests pass.

Version of CURL/LibCURL used: 7.61.1

### CODE REVIEW

File: src/lC_common_h.jl [hunk 1]
```diff
@@ -1,705 +1,1256 @@
...
+const CURLUSESSL_ALL = (UInt32)(3)
+const CURLUSESSL_LAST = (UInt32)(4)
+# end enum ANONYMOUS_9
+
+const CURLFTPSSL_NONE = CURLUSESSL_NONE
+const CURLFTPSSL_TRY = CURLUSESSL_TRY
+const CURLFTPSSL_CONTROL = CURLUSESSL_CONTROL
+const CURLFTPSSL_ALL = CURLUSESSL_ALL
+const CURLFTPSSL_LAST = CURLUSESSL_LAST
+const curl_usessl = Void
```
Comment (omus) [inline src/lC_common_h.jl]: `Void` no longer exists on Julia 1.0. This looks like Clang.jl needs to switch to using `Cvoid`

Comment (nicoleepp) [reply src/lC_common_h.jl]: Looks like that was done here: `https://github.com/ihnorton/Clang.jl/commit/1ec7d258f60725fd25c6691e2806e6b5528ce63b` I will try again using master (was on a release).

File: src/lC_common_h.jl [hunk 2]
```diff
@@ -1,705 +1,1253 @@
...
-  quote
-    const $(esc(fake_t)) = $(esc(real_t))
-  end
-end
+using Compat
+
+const CURL_SOCKET_BAD = -1
+
+# begin enum ANONYMOUS_1
+const ANONYMOUS_1 = UInt32
```
Comment (nicoleepp) [inline src/lC_common_h.jl]: Not sure why these ANONYMOUS constants were generated or if they are needed.

Comment (nicoleepp) [reply src/lC_common_h.jl]: I think they come from badly parsed enums. I am currently going through them all to update them.

**Comment by nicoleepp** [comment]
There is one constant I'm not sure what to do with. It gets generated as 

```julia
const CURL_ZERO_TERMINATED               = ((size_t) -1)
```
but this errors in julia:
```julia
julia> Csize_t(-1)
ERROR: InexactError: check_top_bit(Int64, -1)
Stacktrace:
 [1] throw_inexacterror(::Symbol, ::Any, ::Int64) at ./boot.jl:567
 [2] check_top_bit at ./boot.jl:581 [inlined]
 [3] toUInt64 at ./boot.jl:692 [inlined]
 [4] UInt64(::Int64) at ./boot.jl:722
 [5] top-level scope at none:0
```
I have left the constant defined for now as:

```julia
const CURL_ZERO_TERMINATED               = -1
```

**Comment by nicoleepp** [comment]
My other concern is that some things that used to be `Union{}` are now just `Cvoid`, and constants that were Ints are now specified as `UInt32`.

**Comment by omus** [comment]
The C code probably wants to do:
```julia
julia> reinterpret(Csize_t, -1)
0xffffffffffffffff
```

File: gen/generate.jl [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+# Set these to correspond to your local filesystem's curl and clang include paths
+const CURL_PATH = "/usr/local/opt/curl/include/curl"
+const CLANG_INCLUDES = [
+    "/usr/local/opt/llvm/include",
+]
+
+const SRC_DIR = abspath(@__DIR__, "..", "src")
+
+headers = map(x-> joinpath(CURL_PATH, x), [ "curl.h", "easy.h", "multi.h"])
```
Comment (omus) [inline gen/generate.jl]: Some whitespace issues

File: gen/generate.jl [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+context = wrap_c.init(;
+    headers = headers,
+    clang_args = String[],
+    common_file = joinpath(SRC_DIR, "lC_common_h.jl"),
+    clang_includes = CLANG_INCLUDES,
+    clang_diagnostics = true,
+    header_wrapped = (top_header, cursor_header) -> in(cursor_header, headers),
+    header_library = header -> "libcurl",
+    header_outputfile = header -> joinpath(SRC_DIR, "lC_") * replace(last(split(header, "/")), ".", "_")  * ".jl",
```
Comment (omus) [inline gen/generate.jl]: Probably should be `replace(basename(header), ".", "_")`

File: gen/generate.jl [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    # Generate define constants
+    open(joinpath(SRC_DIR, "lC_defines_h.jl"), "w+") do defines_file
+        println(defines_file, "#   Generating #define constants")
+
+        hashdefs = split(read(`gcc -E -dD -P $(joinpath(CURL_PATH, "curl.h"))`, String), "\n")
+
+        for line in hashdefs
+            m = match(r"^\s*#define\s+CURL(\w+)\s+(.+)", line)
+
+            if (m != nothing)
```
Comment (omus) [inline gen/generate.jl]:
```julia
if m !== nothing
```

File: gen/generate.jl (hunk 4 above)
Comment (ararslan) [reply gen/generate.jl]: Or rather, `replace(basename(header), "." => "_")` for 0.7/1.0 support. This is available from Compat for 0.6.

**Comment by nicoleepp** [comment]
I removed the `c` and `ctypedef` macros because we would have to manually go through the generated code and change the format to use them, but, if it makes for easier reviewing, I can add them back and re-format the definitions.

**Comment by omus** [comment]
@nicoleepp I assume you're good with squashing the commits?

**Comment by omus** [comment]
[Travis also passed](https://travis-ci.org/JuliaWeb/LibCURL.jl/builds/426838219) but seems to have not linked back to this PR

**Comment by nicoleepp** [comment]
I'm good with squashing

**Comment by ararslan** [comment]
Nice work here!

**Pull request merged**